src/com/xilinx/rapidwright/util/JobQueue.java

```diff
@@ -80,33 +80,52 @@ public boolean addRunningJob(Job j) {
+        return running.size();
+    }
+
+    private Map<JobState, List<Job>> getJobsByState() {
+        return running.stream().collect(Collectors.groupingBy(Job::getJobState, () -> new EnumMap<>(JobState.class), Collectors.toList()));
+    }
+
+    public boolean refreshQueue() {
+        return refreshQueue(isLSFAvailable() ? JobQueue.MAX_LSF_CONCURRENT_JOBS : JobQueue.MAX_LOCAL_CONCURRENT_JOBS);
+    }
+    public boolean refreshQueue(int maxNumRunningJobs) {
+        return refreshQueue(maxNumRunningJobs, getJobsByState());
+    }
+
+    private boolean refreshQueue(int maxNumRunningJobs, final Map<JobState, List<Job>> jobsByState) {
+        final List<Job> exited = jobsByState.get(JobState.EXITED);
+        if (exited != null) {
+            for (Job job : exited) {
+                running.remove(job);
+                finished.add(job);
+            }
+            //Removing from map so they don't show up in our printout
+            jobsByState.remove(JobState.EXITED);
+        }
+        boolean launched = false;
+        while (!waitingToRun.isEmpty() && maxNumRunningJobs > running.size()) {
+            Job j = waitingToRun.poll();
+            long pid = j.launchJob();
+            running.add(j);
+            if (printJobStart) {
+                System.out.println("Running job [" + pid + "] " + j.getCommand() + " in " + j.getRunDir());
+            }
+            launched = true;
+        }
+        return launched;
+    }
+
     public boolean runAllToCompletion() {
         return runAllToCompletion(isLSFAvailable() ? JobQueue.MAX_LSF_CONCURRENT_JOBS : JobQueue.MAX_LOCAL_CONCURRENT_JOBS);
     }
 
     public boolean runAllToCompletion(int maxNumRunningJobs) {
         while (!waitingToRun.isEmpty() || !running.isEmpty()) {
-
-            final Map<JobState, List<Job>> jobsByState = running.stream().collect(Collectors.groupingBy(Job::getJobState, ()->new EnumMap<>(JobState.class), Collectors.toList()));
-            final List<Job> exited = jobsByState.get(JobState.EXITED);
-            if (exited != null) {
-                for (Job job : exited) {
-                    running.remove(job);
-                    finished.add(job);
-                }
-                //Removing from map so they don't show up in our printout
-                jobsByState.remove(JobState.EXITED);
-            }
-            boolean launched = false;
-            while (!waitingToRun.isEmpty() && maxNumRunningJobs > running.size()) {
-                Job j = waitingToRun.poll();
-                long pid = j.launchJob();
-                running.add(j);
-                if (printJobStart) {
-                    System.out.println("Running job [" + pid + "] " + j.getCommand() + " in " + j.getRunDir());
-                }
-                launched = true;
-            }
+            final Map<JobState, List<Job>> jobsByState = getJobsByState();
+            boolean launched = refreshQueue(maxNumRunningJobs, getJobsByState());
```

Comment:
What was the motivation behind this change?  It seems like it is refactoring the code, but not sure why.  If we want to keep it this way, can you create a test to show that the changes are correct?

Reply:
If the LSF job queue gets too large (thousands of jobs), then it seems to get stuck. This would happen because the jobQueue would only start the first `maxNumRunningJobs` that were created. All subsequently created jobs would just sit in the queue until `runAllToCompletion()` is called. Periodically calling `refreshQueue()` while jobs are being created allows    jobs that are waiting to run a chance to start running and prevent the job queue from becoming backed up too much.

If you look at the diff of jobQueue.java, you will see that the contents of `refreshQueue()` are pulled directly from `runAllToCompletion()` without any changes (besides adding a return statement). Thus, the refactor shouldn't change the functionality of the code at all. However, I can still create a test if you'd like.